src/components/Item/VisualizationItem/Visualization/IframePlugin.js

```diff
@@ -0,0 +1,106 @@
+import { useD2 } from '@dhis2/app-runtime-adapter-d2'
+import postRobot from '@krakenjs/post-robot'
+import PropTypes from 'prop-types'
+import React, {
+    Suspense,
+    useRef,
+    useCallback,
+    useState,
+    useEffect,
+} from 'react'
+import { itemTypeMap } from '../../../../modules/itemTypes.js'
+import { getPluginOverrides } from '../../../../modules/localStorage.js'
+import { useUserSettings } from '../../../UserSettingsProvider.js'
+import classes from './styles/DataVisualizerPlugin.module.css'
+import VisualizationErrorMessage from './VisualizationErrorMessage.js'
+
+const IframePlugin = ({
+    filterVersion,
+    item,
+    style,
+    visualization,
+    dashboardMode,
+}) => {
+    const { d2 } = useD2()
+    // TODO replace d2 with useCachedDataQuery
+    const apps = d2.system.installedApps
+    const { userSettings } = useUserSettings()
+    const iframeRef = useRef()
+    const [error, setError] = useState(false)
+
+    const onError = () => setError(true)
+
+    useEffect(() => {
+        const listener = postRobot.on(
+            'getProps',
+            // listen for messages coming only from the iframe rendered by this component
+            { window: iframeRef.current.contentWindow },
+            () => ({
+                isVisualizationLoaded: true,
+                nameProp: userSettings.displayProperty,
+                visualization,
+                onError,
+            })
+        )
+
+        return () => listener.cancel()
+    }, [visualization, userSettings.displayProperty])
+
+    useEffect(() => {
+        setError(false)
+    }, [filterVersion, visualization.type])
+
+    const getIframeSrc = useCallback(() => {
+        // 1. check if there is an override for the plugin
+        const pluginOverrides = getPluginOverrides()
+
+        if (pluginOverrides && pluginOverrides[item.type]) {
+            return pluginOverrides[item.type]
+        }
+
+        const appKey = itemTypeMap[item.type].appKey
+
+        // 2. check if there is an installed app for the item type
+        // and use its plugin launch URL
+        if (appKey) {
+            const appDetails = apps.find((app) => app.key === appKey)
+
+            return appDetails?.pluginLaunchUrl
+        }
+
+        return // XXX
+    }, [item.type, apps])
+
+    if (error) {
+        return (
+            <div style={style}>
+                <VisualizationErrorMessage
+                    item={item}
+                    dashboardMode={dashboardMode}
+                />
+            </div>
+        )
+    }
+
+    return (
+        <Suspense fallback={<div />}>
```

Comment:
Will there be a loading spinner like the other visualizations?

Reply:
I think the plugin itself might already have a spinner.
But I don't think we need `Suspense` here, the iframe tag can be rendered right away, the problem is that until the plugin is mounted and got all the props it needs, nothing will be rendered inside the iframe.
This might not be a problem as it does not even depend on network, but there might be other reasons why the iframe content takes time before rendering anything.

Reply:
Removed `Suspense`

Reply:
Clarification:
the new `PluginWrapper` components in the various apps show a loading spinner while the communication between dashboard app and iframe plugin content happen.